### DISCUSSION TITLE
Git/GitHub-like stuff

### DISCUSSION THREAD

**Discussion by Pawper** [beakerbrowser/beaker#1662] [General]
Beaker currently has a forking, copying & a diff / merge screen which if used for development is very dependent upon manual communication. So I have some ideas:
- Could Beaker list others' forks of your hyperdrive in the diff / merge screen by crawling hypderdrives for index.json files with the `.forkOf:` matching the drive's URL?

- Could a hyperdrive's index.json have a voluntary `author:URL` key/value pair? When a user creates, forks or copies a hyperdrive, Beaker could:
  - Prompt the user: **Add your profile address to the manifest?** In the new hyperdrive's index.json, this adds the `author:URL` key value pair, where URL is the Beaker installation's profile URL. Fork author photo/name will show in the diff / merge UI's dropdown
    - If yes, **Validate in your profile hyperdrive?** This adds an object for the fork in the validation.json in one's profile hyperdrive. Validation symbol ✔ will show in the diff / merge UI's dropdown, or a warning symbol ⚠ if not validated
    - If no, it's essentially anonymous, like how hyperdrives are now.

- Could a hyperdrive's index.json have a `pullFlag:` key? This is like in real life when you raise the flag on a curbside mailbox. When Beaker crawls hypdrives for forks, it can also look for this. Then when the fork is ready, you can raise the pull flag and it will show in the diff / merge UI's dropdown 🚩
  - In settings we can set Beaker to show the icon as a notification in this area:
![image](https://user-images.githubusercontent.com/22269691/82743137-f3e88080-9d1b-11ea-8fe0-cb9b0053c79e.png)

- Could a hyperdrive have a `bugDrive: yes` in its manifest and bugs.json file? Beaker could add UI for filing a bug. It create a fork of the hyperdrive and prompts the user to add a markdown file. The markdown file's name is added to bugs.json and the `pullFlag` mentioned above is set to bug 🐛

Default for the above items could be set in Beaker settings.

The only thing I am not sure how to resolve is versioning with commit messages. This used to be in dat. How would it look in hyper? Could we just use git in the terminal? I'd love to have some shortcuts while editing to quickly stage changes & add a commit message. I currently use Github Desktop on my workstation, because I find it's so much quicker.

**Comment by da2x** [comment]
The normal way to solve this is to provide an email address and a notice in README and CONTRIBUTING instructing people to “Send your bug reports and git-patches to some-project@example.com!”

You can build a bug tracker and a pull-request queue system (a place for people to submit git-patches) by combining https://docs.beakerbrowser.com/apis/beaker.peersockets and https://beaker.dev/docs/tutorials/self-modifying-site/ You’d need to maintain one constantly running Beaker instance with the dirve’s private key for this to work. But it’s absolutely doable.

Have a look at hyper://e219fb97612d4a2b6bee1f5374becbd66de0217c8b831ddbd2cfa66621a9868f/ for a cool example.

> Could Beaker list others' forks of your hyperdrive in the diff / merge screen by crawling hypderdrives for index.json files with the .forkOf: matching the drive's URL?

There is no way to crawl for Hyperdrives without randomly guessing keys. That potential number of drives is absolutely enormous. You’re unlikely to ever guess a single active drive hash, yet alone one that hosts data you’d be interested in.

With BitTorrent, you can monitor the DHT for torrent hashes. However, Dat uses a DHT hash derived from the drive’s real drive hash. So you can see that peers are queuing the DHT about *something* but you don’t know the real drive hash. You need the drive hash to request that peers send you any data.

**Comment by pfrazee** [comment]
I think there are some interesting ideas here and I do think building more into the collab tools would be really nice
